### ISSUE TITLE
Changing ON DELETE & ON UPDATE constraints not propagated

### ISSUE THREAD

**Issue by pndiku** [perseas/Pyrseas#134]
If you drop a foreign key and recreate it with different ON DELETE and ON CASCADE parameter, the YAML is correctly generated by dbtoyaml.

However, running yamltodb on the secondary database does not drop and recreate these constraints.

**Comment by jmafc** [comment]
Adding and dropping of foreign keys was done recenty (see for example change 3b8beb7 by @vayerx ). Are you testing with HEAD of the master branch?

**Comment by pndiku** [comment]
I can't use HEAD because it introduces another bug, namely:

DB1 has privileges, but I do a ``dbtoyaml -O -x`` to skip the privileges because DB2 has different privileges.

When I run a yamltodb on DB2 I get an error:

```
Traceback (most recent call last):
  File "/usr/bin/yamltodb", line 9, in <module>
    load_entry_point('Pyrseas==0.7.2', 'console_scripts', 'yamltodb')()
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/yamltodb.py", line 47, in main
    stmts = db.diff_map(inmap)
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/database.py", line 354, in diff_map
    stmts = self.db.schemas.diff_map(self.ndb.schemas)
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/schema.py", line 347, in diff_map
    stmts.append(self[sch].diff_map(insch))
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/__init__.py", line 370, in diff_map
    stmts.append(self.diff_privileges(inobj))
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/__init__.py", line 380, in diff_privileges
    return diff_privs(self, self.privileges, inobj, inobj.privileges)
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/privileges.py", line 182, in diff_privs
    (usr, privcodes, grantor) = _split_privs(privspec)
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/privileges.py", line 21, in _split_privs
    (usr, prvgrant) = privspec.split('=')
ValueError: need more than 1 value to unpack
```

**Comment by jmafc** [comment]
Yes, unfortunately, `yamltodb` was never updated to have an equivalent of `dbtoyaml`'s `-x` option, i.e., skip over comparison of privileges. OTOH, it would be difficult to integrate @vayerx 's changes into the `r0.7` branch. So I'm not sure how we could work this out.

I see you're using Python 2.7. I presume that you were originally using Pyrseas 0.7.2, like installed from PyPI, but that the above traceback comes from using the HEAD of the github master branch? [Note to self: we should probably change master's `setup.py` and `pyrseas/__init__.py` to use a version like '0.8.dev0'). For testing purposes, it may also help us if you tell us the PG version(s) you're using.

Also, please take a look at https://github.com/perseas/Pyrseas/blob/master/tests/dbobject/test_constraint.py#L724, in particular the `stmts` at the beginning of the function and the last two `assertEqual` statements at the end of the function, and describe what you would change to come up with a testcase to match your specific situation.

**Comment by pndiku** [comment]
Yes, I'm using the HEAD branch and even when I do dbtoyaml _without_ the `-x` option, I still get the same error about privileges. When I look in the YAML I see things like this:

```
  owner: pndiku
  sequence outcome_agency_outcome_agency_id_seq:
    cache_value: 1
    increment_by: 1
    max_value: null
    min_value: null
    owner: pndiku
    owner_column: outcome_agency_id
    owner_table: outcome_agency
    privileges: []
    start_value: 1
```

The privileges field is empty but yamltodb still says `need more than 1 value to unpack`

**Comment by jmafc** [comment]
That looks like the side effect of some other change, because the HEAD of `r0.7` doesn't give any of those "privileges: []" lines for sequences (on the same database). I'll try to take a more detailed look later on today.

**Comment by jmafc** [comment]
Thanks to `git bisect` I now know the "privileges: []" come from my change 5da6b0c. Later I'll try to fix that in HEAD.

**Comment by jmafc** [comment]
HEAD of master now includes a fix for those non-specified privileges of sequences showing up as empty lists.  Please try to test with the current HEAD to see if that resolves your original problem.

**Comment by pndiku** [comment]
Unfortunately not. It's still happening:

```
Traceback (most recent call last):
  File "/usr/bin/yamltodb", line 9, in <module>
    load_entry_point('Pyrseas==0.7.2', 'console_scripts', 'yamltodb')()
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/yamltodb.py", line 47, in main
    stmts = db.diff_map(inmap)
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/database.py", line 354, in diff_map
    stmts = self.db.schemas.diff_map(self.ndb.schemas)
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/schema.py", line 347, in diff_map
    stmts.append(self[sch].diff_map(insch))
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/__init__.py", line 370, in diff_map
    stmts.append(self.diff_privileges(inobj))
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/__init__.py", line 380, in diff_privileges
    return diff_privs(self, self.privileges, inobj, inobj.privileges)
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/privileges.py", line 182, in diff_privs
    (usr, privcodes, grantor) = _split_privs(privspec)
  File "/usr/lib/python2.7/site-packages/Pyrseas-0.7.2-py2.7.egg/pyrseas/dbobject/privileges.py", line 21, in _split_privs
    (usr, prvgrant) = privspec.split('=')
ValueError: need more than 1 value to unpack
```

This is from doing a `dbtoyaml` both with and without the `-x` flag, then doing a yamltodb on the same database.

**Comment by pndiku** [comment]
Interestingly, in the YAML file, I only see one "privileges:" line - under one of my Pl/PgSQL functions

**Comment by jmafc** [comment]
It's somewhat unclear. You did a `dbtoyaml` with and without the `-x` flag.  Presumably in the first case you didn't get any "privileges:" lines, right? And the error comes from processing the second output file, i.e., the one with a single "privilege:" line?  But I think what you want is to use the first output, i.e., without privileges since you'll eventually test against another database that has different privileges.

Note that I first want to verify that the ON DELETE/ON CASCADE issue is or is not present in the master branch (because our testcases says it shouldn't be).

**Comment by pndiku** [comment]
I've created two brand new databases that are duplicates with 2 tables - testdb and testdb2. 

YAML for testdb (`yamltodb -O -x testdb`) gives this:
```
extension plpgsql:
  description: PL/pgSQL procedural language
  schema: pg_catalog
  version: '1.0'
schema public:
  description: standard public schema
  table test:
    columns:
    - col_a:
        not_null: true
        type: integer
    - col_b:
        type: text
    primary_key:
      test_pkey:
        columns:
        - col_a
  table test2:
    columns:
    - col_a:
        type: integer
    - col_b:
        type: integer
    foreign_keys:
      test2_fk2:
        columns:
        - col_a
        on_delete: cascade
        on_update: cascade
        references:
          columns:
          - col_a
          schema: public
          table: test
```

When I run a `yamltodb -u testdb2 testdb.yaml`, I get the privileges error as previously explained. 

Second test I did was to drop testdb2. Then recreate it.

Now, running `yamltodb -u testdb2 testdb.yaml` works.

However, if I _immediately_ run `yamltodb -u testdb2 testdb.yaml` AGAIN, I get the privileges error.

So, basically I can't even test the ON DELETE, ON CASCADE issue.

**Comment by jmafc** [comment]
OK, I've reproduced the problem, but as you point out, it only happens on the second run.  The problem appears to be that the first `yamltodb` issues the following commands:

```sql
BEGIN;
REVOKE ALL ON SCHEMA public FROM postgres;

REVOKE ALL ON SCHEMA public FROM PUBLIC;

COMMIT;
```
The `_split_privs` error on the second run seems to come from having to deal with those missing privileges. If I issue (on the second database)

```sql
GRANT ALL ON SCHEMA public TO postgres;
GRANT ALL ON SCHEMA public TO PUBLIC;
```
and then run `yamltodb -1` with the same YAML file (against the second database), the REVOKE's reappear but there is no error.

I'll have to look at what is causing these REVOKEs, sometime later.

**Comment by jmafc** [comment]
Peter, change dd2cb80 (just submitted) gets rid of the `_split_privs` error on the second run. Please give it a try. (Note: I still have to check the REVOKEs on the first run).

**Comment by pndiku** [comment]
Perfect! I'll close this unless you want to keep it open for the REVOKE issue.

Thanks so much Joe!
